bundles/admin/admin-hierarchical-layerlist/instance.js

```diff
@@ -0,0 +1,364 @@
+         * @method  _addSubgroupSubgroupTools
+         * @private
+         */
+        _addSubgroupSubgroupTools: function() {
+            var me = this;
+            // Add layer add tools
+            me.service.addSubgroupSubgroupTool('add-layer', function(tool, groupId, parentId) {
+                me.layer.showLayerAddPopup(tool, null, groupId);
+            }, {
+                cls: 'add-layer',
+                tooltip: me.locale('tooltips.addLayer')
+            });
+
+            // Add edit tool to adding groups
+            me.service.addSubgroupSubgroupTool('edit-subgroup-subgroup', function(tool, groupId, parentId) {
+                var group = me.layerService.getAllLayerGroups(groupId);
+                var options = {
+                    locale: group.name,
+                    selectable: group.selectable
+                };
+                options.type = 'subgroup-subgroup';
+
+                var popupConf = me.group.getGroupAddingPopupConf(tool, groupId, parentId, options);
+                var popup = popupConf.popup;
+                var message = popupConf.message;
+                popupConf.popup.show(me.locale('groupTitles.editSubgroup'), message, popupConf.buttons);
+                popupConf.popup.makeModal();
+            }, {
+                cls: 'edit-subgroup-subgroup',
+                tooltip: me.locale('tooltips.editSubgroup')
+            });
+        },
+        /**
+         * Add layertree options
+         * @method  _addOptions
+         * @private
+         */
+        _addOptions: function() {
+            var me = this;
+
+            // Add Drag & drop plugin
+            me.service.addLayerlistOption('plugins', ['checkbox', 'changed', 'wholerow', 'types', 'search', 'state', 'conditionalselect', 'dnd'], false);
+            me.service.addLayerlistOption('dnd', {
+                use_html5: true,
+                inside_pos: 'last'
+            });
+        },
+
+        /**
+         * Add layer tools
+         * @method  _addLayerTools
+         * @private
+         */
+        _addLayerTools: function() {
+            var me = this;
+            // Add edit tool for layers
+            me.service.addLayerTool('edit-layer', function(tool, groupId, layerId) {
+
+                me.layer.showLayerAddPopup(tool, layerId, groupId);
+            }, {
+                cls: 'edit-layer',
+                tooltip: me.locale('tooltips.editLayer')
+            });
+        },
+
+        /**
+         * Assigns event handlers to the admin hierarchical layer list.
+         */
+        _addEventHandlers: function() {
+            var me = this;
+            jQuery(document).on("dnd_stop.vakata", function(event, data) {
+                //If the drag target group is not open, we have to open it.
+                //Otherwise we can't get the necessary information of the drag operation.
+                var targetGroup = data.data.origin.get_node(jQuery(data.event.target).prop("id").split("_")[0]);
```

Comment:
What's the structure of "data" here? What's the difference between event and data.event? Are dnd-targets restricted to elements that always have id (and with a underscore to split from)? Can you add a comment about expected content of data and event target. Is this the same dnd we use in layerselection2 or another lib?

Reply:
dnd_stop.vakata is jstree event and not used layerselection2 bundle.
Jstree bind this event in document (not jstree element) so need be there, handles drag-n drop.
See: https://www.jstree.com/api/#/?f=dnd_stop.vakata 

I add comments.